### PR TITLE
colencoding: reuse scratch space when key decoding bytes and decimals

### PR DESCRIPTION
**Pull request by yuzefovich**
When we're key decoding bytes-like columns, we need to use the scratch
byte slice to decode into (in case of decimals, we might need the space
temporarily). Previously, we would always allocate a new byte slice only
to deep copy it later when calling `coldata.Bytes.Set`. This commit
teaches the cFetcher and the relevant decoding methods to reuse the same
scratch space which should reduce the memory allocations.

One notable change is that now when we're calling
`DecodeBytesAscending`, we have to make sure to perform a deep copy so
that it is safe to reuse the returned value as the scratch space in the
future.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
Okay, this doesn't just work, yet.

**Comment by yuzefovich** [comment]
I think it should now work. RFAL.

**Review by mgartner — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70734)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/colencoding/key_encoding.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70734#-MkcXFvi13KbKNJY-kTU:-MkcXFvi13KbKNJY-kTV:b-e9bshf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d78f0aa0ae5a9f0b4f759fa0dc329b2c4e643b2e/pkg/sql/colencoding/key_encoding.go#L294)):*
> ```Go
> 		}
> 		// Note that since we're about to perform a deep copy, it'll be ok to
> 		// return the scratch slice to be reused by the caller.
> ```

I don't understand this comment. Since we've already deep-copied, why does it matter that `Set` deep-copies or not?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70734)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[pkg/sql/colencoding/key_encoding.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70734#-MkcXFvi13KbKNJY-kTU:-MkcXslNE_aZuPTnQi6p:b-60p2e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d78f0aa0ae5a9f0b4f759fa0dc329b2c4e643b2e/pkg/sql/colencoding/key_encoding.go#L294)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't understand this comment. Since we've already deep-copied, why does it matter that `Set` deep-copies or not?
</blockquote></details>

There are two concerns at play: one is about not referencing the memory used by `key` (to allow for GC to run on the batch response), another one is making sure it is ok to return `scratch` in such a manner that if `scratch` is later modified, then `vec.Bytes().Get(idx)` will still return the correct result. The first concern is addressed by making a deep-copy when decoding, the second concern is addressed automatically by `vec.Bytes().Set()` which performs the deep copy. The comment is about the second concern.

Do you think it's worth extending the comment? Or maybe removing it to reduce possible confusion?

**Review by mgartner — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70734#-:-MkhScY10mZnfN76ZI1H:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/colencoding/key_encoding.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70734#-MkcXFvi13KbKNJY-kTU:-MkhS8Mb3Xcoq0t25XEq:b-apdakt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d78f0aa0ae5a9f0b4f759fa0dc329b2c4e643b2e/pkg/sql/colencoding/key_encoding.go#L294)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

There are two concerns at play: one is about not referencing the memory used by `key` (to allow for GC to run on the batch response), another one is making sure it is ok to return `scratch` in such a manner that if `scratch` is later modified, then `vec.Bytes().Get(idx)` will still return the correct result. The first concern is addressed by making a deep-copy when decoding, the second concern is addressed automatically by `vec.Bytes().Set()` which performs the deep copy. The comment is about the second concern.

Do you think it's worth extending the comment? Or maybe removing it to reduce possible confusion?
</blockquote></details>

Got it. I think maybe just reword it to something like: "Set() performs a deep copy, so it is safe to return the scratch slice to the caller. Any modifications to the scratch slice made by the caller will not affect the value in the vector".

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70734#-:-MkhUC5L4Rl3vG331d2M:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)
